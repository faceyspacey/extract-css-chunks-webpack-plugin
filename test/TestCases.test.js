import fs from 'fs';
import path from 'path';

import webpack from 'webpack';
import execa from 'execa';

describe('TestCases', () => {
  const casesDirectory = path.resolve(__dirname, 'cases');
  const outputDirectory = path.resolve(__dirname, 'js');
  for (const directory of fs.readdirSync(casesDirectory)) {
    if (!/^(\.|_)/.test(directory)) {
      // eslint-disable-next-line no-loop-func
      it(`${directory} should compile to the expected result`, (done) => {
        const directoryForCase = path.resolve(casesDirectory, directory);
        const outputDirectoryForCase = path.resolve(outputDirectory, directory);
        // eslint-disable-next-line import/no-dynamic-require, global-require
        const webpackConfig = require(path.resolve(
          directoryForCase,
          'webpack.config.js'
        ));
        for (const config of [].concat(webpackConfig)) {
          Object.assign(
            config,
            {
              mode: 'none',
              context: directoryForCase,
              output: Object.assign(
                {
                  path: outputDirectoryForCase,
                },
                config.output
              ),
            },
            config
          );
        }
        webpack(webpackConfig, (err, stats) => {
          if (err) {
            done(err);
            return;
          }
          done();
          // eslint-disable-next-line no-console
          console.log(
            stats.toString({
              context: path.resolve(__dirname, '..'),
              chunks: true,
              chunkModules: true,
              modules: false,
            })
          );
          if (stats.hasErrors()) {
            done(
              new Error(
                stats.toString({
                  context: path.resolve(__dirname, '..'),
                  errorDetails: true,
                })
              )
            );
            return;
          }
          const expectedDirectory = path.resolve(directoryForCase, 'expected');
          for (const file of fs.readdirSync(expectedDirectory)) {
            const content = fs.readFileSync(
              path.resolve(expectedDirectory, file),
              'utf-8'
            );
            const actualContent = fs.readFileSync(
              path.resolve(outputDirectoryForCase, file),
              'utf-8'
            );
            expect(actualContent).toEqual(content);
          }
          done();
        });
      }, 10000);
    }
  }
});

describe('HMR', () => {
  it('matches snapshot', () => {
    const hmr = fs
      .readFileSync(path.join(__dirname, '../src/hmr/hotModuleReplacement.js'))
      .toString();

    expect(hmr).toMatchSnapshot();
  });

  it('is es5 only', () => {
    const { stderr } = execa.shellSync(
      'npx es-check es5 src/hmr/hotModuleReplacement.js'
    );

    expect(
      stderr.indexOf('there were no ES version matching errors') > -1
    ).toBe(true);
  });
});
