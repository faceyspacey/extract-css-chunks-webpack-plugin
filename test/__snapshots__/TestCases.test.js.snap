// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HMR matches snapshot 1`] = `
"/* global document, window */
/* eslint func-names: 0 */
/* eslint no-var: 0 */
/* eslint vars-on-top: 0 */
/* eslint prefer-arrow-func: 0 */
/* eslint prefer-rest-params: 0 */
/* eslint prefer-arrow-callback: 0 */
/* eslint prefer-template: 0 */

var normalizeUrl = require('normalize-url');

var srcByModuleId = Object.create(null);

var noDocument = typeof document === 'undefined';

var forEach = Array.prototype.forEach;

function debounce(fn, time) {
  var timeout = 0;

  // eslint-disable-next-line func-names
  return function() {
    var self = this;
    var args = arguments;

    // eslint-disable-next-line prefer-rest-params
    var functionCall = function functionCall() {
      return fn.apply(self, args);
    };

    clearTimeout(timeout);
    timeout = setTimeout(functionCall, time);
  };
}

function noop() {}

function getCurrentScriptUrl(moduleId) {
  var src = srcByModuleId[moduleId];

  if (!src) {
    if (document.currentScript) {
      src = document.currentScript.src;
    } else {
      var scripts = document.getElementsByTagName('script');
      var lastScriptTag = scripts[scripts.length - 1];

      if (lastScriptTag) {
        src = lastScriptTag.src;
      }
    }
    srcByModuleId[moduleId] = src;
  }

  return function(fileMap) {
    if (!src) {
      return null;
    }
    var splitResult = src.split(/([^\\\\\\\\/]+)\\\\.js$/);
    var filename = splitResult && splitResult[1];
    if (!filename) {
      return [src.replace('.js', '.css')];
    }
    if (!fileMap) {
      return [src.replace('.js', '.css')];
    }
    return fileMap.split(',').map(function(mapRule) {
      var reg = new RegExp(filename + '\\\\\\\\.js$', 'g');
      return normalizeUrl(
        src.replace(reg, mapRule.replace(/{fileName}/g, filename) + '.css'),
        { stripWWW: false }
      );
    });
  };
}

function updateCss(el, url) {
  if (!url) {
    url = el.href.split('?')[0];
  }
  if (el.isLoaded === false) {
    // We seem to be about to replace a css link that hasn't loaded yet.
    // We're probably changing the same file more than once.
    return;
  }
  if (!url || !(url.indexOf('.css') > -1)) return;

  el.visited = true;
  var newEl = el.cloneNode(); // eslint-disable-line vars-on-top

  newEl.isLoaded = false;

  newEl.addEventListener('load', function() {
    newEl.isLoaded = true;
    el.parentNode.removeChild(el);
  });

  newEl.addEventListener('error', function() {
    newEl.isLoaded = true;
    el.parentNode.removeChild(el);
  });

  newEl.href = url + '?' + Date.now();
  el.parentNode.appendChild(newEl);
}

function getReloadUrl(href, src) {
  var ret;
  href = normalizeUrl(href, { stripWWW: false });
  // eslint-disable-next-line array-callback-return
  src.some(function(url) {
    if (href.indexOf(src) > -1) {
      ret = url;
    }
  });
  return ret;
}

function reloadStyle(src) {
  var elements = document.querySelectorAll('link');
  var loaded = false;

  forEach.call(elements, function(el) {
    var url = getReloadUrl(el.href, src);

    if (el.visited === true) return;

    if (url) {
      updateCss(el, url);
      loaded = true;
    }
  });

  return loaded;
}

function reloadAll() {
  var elements = document.querySelectorAll('link');
  forEach.call(elements, function(el) {
    if (el.visited === true) return;
    updateCss(el);
  });
}

module.exports = function(moduleId, options) {
  if (noDocument) {
    console.log('no window.document found, will not HMR CSS');
    return noop;
  }

  // eslint-disable-next-line vars-on-top
  var getScriptSrc = getCurrentScriptUrl(moduleId);

  function update() {
    var src = getScriptSrc(options.filename);
    var reloaded = reloadStyle(src);

    if (options.locals) {
      console.log('[HMR] Detected local css modules. Reload all css');
      reloadAll();
      return;
    }

    if (reloaded && !options.reloadAll) {
      console.log('[HMR] css reload %s', src.join(' '));
    } else {
      console.log('[HMR] Reload all css');
      reloadAll();
    }
  }

  return debounce(update, 50);
};
"
`;
